### PR TITLE
2730 Create ESS API Endpoint /ProductData Sincedatetime Stub

### PR DESCRIPTION
**Pull request by shailen14599**
Code changes 
Unit tests
Acceptance tests

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Request/ProductDataSinceDateTimeRequest.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using Swashbuckle.AspNetCore.Annotations;
+using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.ExchangeSetService.Common.Models.Request
+{
+    [ExcludeFromCodeCoverage]
```
Comment (WorkRD) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Request/ProductDataSinceDateTimeRequest.cs:6]:
```suggestion
    [ExcludeFromCodeCoverage]
```

Although I do not expect unit tests for the POCOs we should not exclude them by default sine these are used as the input data for our controllers. they should get implicit coverage

which lets us identify if there are any edge cases we might have missed

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Extensions/CallbackUriExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+namespace UKHO.ExchangeSetService.API.Extensions
+{
+    public static class CallbackUriExtensions
+    {
+        public static bool IsValidCallbackUri(this string callbackUri)
+        {
+            try
+            {
+                Uri baseUri = new Uri(callbackUri);
+                return (baseUri.Scheme == Uri.UriSchemeHttp || baseUri.Scheme == Uri.UriSchemeHttps);
```
Comment (WorkRD) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Extensions/CallbackUriExtensions.cs:12]: This should be raised in your threat modeling.,

We should not be calling HTTP

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        [Test()]
+        public async Task WhenValidateProductDataSinceDateTimeInRequest_ThenCreateProductDataSinceDateTimeReturnSuccess()
+        {
+            A.CallTo(() => fakeIProductDataValidatorService.Validate(A<ProductDataSinceDateTimeRequest>.Ignored))
+                .Returns(new ValidationResult(new List<ValidationFailure>()));
+
+            var result = await service.CreateProductDataSinceDateTime(new ProductDataSinceDateTimeRequest());
+            Assert.IsInstanceOf<ExchangeSetResponse>(result);
```
Comment (WorkRD) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs:70]: Can we throw in an assert of the status code here too?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ProductDataControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        {
+
+            A.CallTo(() => fakeproductDataService.ValidateProductDataSinceDateTime(A<ProductDataSinceDateTimeRequest>.Ignored))
+                 .Returns(new ValidationResult(new List<ValidationFailure>()));
+
+            A.CallTo(() => fakeproductDataService.CreateProductDataSinceDateTime(A<ProductDataSinceDateTimeRequest>.Ignored))
+                .Returns(new ExchangeSetResponse());
+
+            var result = (OkObjectResult)await controller.GetProductDataSinceDateTime("Wed, 21 Oct 2015 07:28:00 GMT", "www.abc.com");
+            Assert.IsInstanceOf<OkObjectResult>(result);
```
Comment (WorkRD) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ProductDataControllerTests.cs:65]: Can we do an assert that the result is the same as the response given from the CreateProductDataSinceDateTime response

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        public async Task WhenICallTheApiWithAValidRFC1123DateTimeAndValidCallbackURL_ThenASuccessStatusIsReturned()
+        {
+            string sincedatetime = "Mon, 01 Mar 2021 00:00:00 GMT";
+            var apiresponse = await ExchangesetApiClient.GetExchangeSetBasedOnDateTimeAsync(sincedatetime, "http://fss.ukho.gov.uk/batch/7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272");
+            Assert.AreEqual(200, (int)apiresponse.StatusCode, $"Exchange Set for datetime is  returned {apiresponse.StatusCode}, instead of the expected 200.");
+
+        }
+
+        [TestCase(0, TestName = "Current DateTime with valid RFC1123 format")]
+        [TestCase(1, TestName = "Future DateTime with valid RFC1123 format")]
```
Comment (WorkRD) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs:55]: this is a great use of the TestName property- thanks

**Review by WorkRD — CHANGES_REQUESTED** [review]
some minor requests

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            var errorMessage = await apiresponse.ReadAsTypeAsync<ErrorDescriptionResponseModel>();
+            Assert.IsTrue(errorMessage.Errors.Any(e => e.Source == "SinceDateTime"));
+            Assert.IsTrue(errorMessage.Errors.Any(e => e.Description == "Query parameter 'SinceDateTime' is required."));
+        }
+
+        [TestCase("fss.ukho.gov.uk", TestName = "Callback URL without http or https")]
+        [TestCase("https:/fss.ukho.gov.uk", TestName = "Callback URL with wrong https parameter")]
+        [TestCase("http:/fss.ukho.gov.uk", TestName = "Callback URL with wrong http parameter")]
+        [TestCase("https://", TestName = "Callback URL with only https parameter")]
+        public async Task WhenICallTheApiWithInvalidCallbackURI_ThenABadRequestStatusIsReturned(string callbackurl)
```
Comment (gregzealley) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs:98]:
```suggestion
        public async Task WhenICallTheApiWithInvalidCallbackURI_ThenABadRequestResponseIsReturned(string callbackurl)
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+using NUnit.Framework;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using UKHO.ExchangeSetService.API.Services;
+using UKHO.ExchangeSetService.API.Validation;
+using UKHO.ExchangeSetService.Common.Models.Request;
+using UKHO.ExchangeSetService.Common.Models.Response;
+
+namespace UKHO.ExchangeSetService.API.UnitTests.Services
```
Comment (gregzealley) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs:12]: The brackets on the NUnit tags are not required, e,g, [TestFixture], [Test]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                .Returns(new ValidationResult(new List<ValidationFailure>
+                            { new ValidationFailure("SinceDateTime", "Provided since date time cannot be a future date.")}));
+
+            var result = await service.ValidateProductDataSinceDateTime(new ProductDataSinceDateTimeRequest());
+            Assert.IsFalse(result.IsValid);
+            Assert.AreEqual("Provided since date time cannot be a future date.", result.Errors.Single().ErrorMessage);
+        }
+
+        [Test()]
+        public async Task WhenCallbackUrlparameterNotValidInRequest_ThenValidateProductDataSinceDateTimeReturnsBadRequest()
```
Comment (gregzealley) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs:52]:
```suggestion
        public async Task WhenCallbackUrlParameterNotValidInRequest_ThenValidateProductDataSinceDateTimeReturnsBadRequest()
```

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductDataSinceDateTimeValidatorTests.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
+﻿using FluentValidation.Results;
+using FluentValidation.TestHelper;
+using NUnit.Framework;
+using System;
+using System.Linq;
+using UKHO.ExchangeSetService.API.Validation;
+using UKHO.ExchangeSetService.Common.Models.Request;
+
+namespace UKHO.ExchangeSetService.API.UnitTests.Validation
```
Comment (gregzealley) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductDataSinceDateTimeValidatorTests.cs:9]: Again with the brackets in the NUnit tags

**Review by gregzealley — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs (hunk 5 above)
Comment (mrunmay14436) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs:55]: Thanks Rob.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Extensions/CallbackUriExtensions.cs (hunk 2 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API/Extensions/CallbackUriExtensions.cs]: Changes done to avoid calling HTTP and it will only allow HTTPS.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ProductDataControllerTests.cs (hunk 4 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Controllers/ProductDataControllerTests.cs:65]: Changes has been done for result is the same as the response given from the CreateProductDataSinceDateTime response.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Response/InternalServerError.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.ExchangeSetService.Common.Models.Response
+{
+    [ExcludeFromCodeCoverage]
+    public class InternalServerError
+    {
+        public string CorrelationId { get; set; }
+        public string Details { get; set; }
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Response/InternalServerError.cs:9]: Kindly change this to public string Detail { get; set; } to match according to API spec.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        private TestConfiguration Config { get; set; }
+
+        [SetUp]
+        public void Setup()
+        {
+            Config = new TestConfiguration();
+            ExchangesetApiClient = new ExchangeSetApiClient(Config.EssBaseAddress);
+        }
+        [Test]
+        public async Task WhenICallTheApiWithAValidRFC1123DateTime_ThenASuccessStatusIsReturned()
```
[the remark this answers is not in the document]
Comment (kajalkundu) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs]: Updated the Test Name as suggested. Thanks Greg.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs (hunk 6 above)
Comment (kajalkundu) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetAfterSpecifiedDateTime.cs]: Updated the Test Name as suggested. Thanks Greg.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs (hunk 7 above)
Comment (WorkRD) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs:12]: Hi Greg, although they are not necessarily required they still are legitimate code, I am not going to block a pull request for it

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Request/ProductDataSinceDateTimeRequest.cs (hunk 1 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Models/Request/ProductDataSinceDateTimeRequest.cs]: We have Removed ExcludeFromCodeCoverage.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs (hunk 7 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs:12]: We have removed [TestFixture] and [Test] brackets.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs (hunk 8 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Services/ProductDataServiceTests.cs]: We have change method naming convention.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductDataSinceDateTimeValidatorTests.cs (hunk 9 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.UnitTests/Validation/ProductDataSinceDateTimeValidatorTests.cs:9]: We have removed brackets in the NUnit tags.

**Review by janvi13603 — APPROVED** [review]

**Review by WorkRD — APPROVED** [review]

**Review by rockydevnet — CHANGES_REQUESTED** [review]
